src/display/annotation_storage.js

```diff
@@ -49,6 +53,9 @@ class AnnotationStorage {
    * @param {Object} value
    */
   setValue(key, value) {
+    if (this._storage.get(key) !== value) {
```

Comment:
The only situation where this won't work is https://github.com/mozilla/pdf.js/blob/b061c300b48531204bc136ce0bc4124a8ba43ef8/src/display/annotation_layer.js#L687 because that class is the only widget that doesn't use `getOrCreateValue` first. You can verify this by opening https://github.com/mozilla/pdf.js/blob/master/test/pdfs/annotation-choice-widget.pdf and without changing anything navigating away from the page, and you'll get the warning.

However, I'll most likely need to change that anyway for #12233, so this should be fine, but merging this patch is therefore blocked on fixing that issue, which I'm working on right now.

Reply:
We're getting down to the wire on Firefox's soft freeze coming up. Since the failing behavior isn't too bad, I'm going to merge this in so we have as small as possible updates in mozilla central.